apollo-federation/src/sources/connect/json_selection/methods.rs

```diff
@@ -0,0 +1,2151 @@
+use super::VarsWithPathsMap;
+
+type ArrowMethod = fn(
+    // Method name
+    &str,
+    // Arguments passed to this method
+    &Option<MethodArgs>,
+    // The JSON input value (data)
+    &JSON,
+    // The variables
+    &VarsWithPathsMap,
+    // The input_path (may contain integers)
+    &InputPath<JSON>,
+    // The rest of the PathList
+    &PathList,
+    // Errors
+    &mut IndexSet<ApplyToError>,
+) -> Option<JSON>;
+
+lazy_static! {
+    // This set controls which ->methods are exposed for use in connector
+    // schemas. Non-public methods are still implemented and tested, but will
+    // not be returned from lookup_arrow_method outside of tests.
+    static ref PUBLIC_ARROW_METHODS: IndexSet<&'static str> = {
+        let mut public_methods = IndexSet::default();
+
+        public_methods.insert("echo");
+        // public_methods.insert("typeof");
+        public_methods.insert("map");
+        // public_methods.insert("eq");
+        public_methods.insert("match");
+        // public_methods.insert("matchIf");
+        // public_methods.insert("match_if");
+        // public_methods.insert("add");
+        // public_methods.insert("sub");
+        // public_methods.insert("mul");
+        // public_methods.insert("div");
+        // public_methods.insert("mod");
+        public_methods.insert("first");
+        public_methods.insert("last");
+        public_methods.insert("slice");
+        public_methods.insert("size");
+        // public_methods.insert("has");
+        // public_methods.insert("get");
+        // public_methods.insert("keys");
+        // public_methods.insert("values");
+        public_methods.insert("entries");
+        // public_methods.insert("not");
+        // public_methods.insert("or");
+        // public_methods.insert("and");
+
+        public_methods
+    };
+
+    // This map registers all the built-in ->methods that are currently
+    // implemented, even the non-public ones that are not included in the
+    // PUBLIC_ARROW_METHODS set.
+    static ref ARROW_METHODS: IndexMap<String, ArrowMethod> = {
+        let mut methods = IndexMap::<String, ArrowMethod>::default();
+
+        // This built-in method returns its first input argument as-is, ignoring
+        // the input data. Useful for embedding literal values, as in
+        // $->echo("give me this string").
+        methods.insert("echo".to_string(), echo_method);
+
+        // Returns the type of the data as a string, e.g. "object", "array",
+        // "string", "number", "boolean", or "null". Note that `typeof null` is
+        // "object" in JavaScript but "null" for our purposes.
+        methods.insert("typeof".to_string(), typeof_method);
+
+        // When invoked against an array, ->map evaluates its first argument
+        // against each element of the array and returns an array of the
+        // results. When invoked against a non-array, ->map evaluates its first
+        // argument against the data and returns the result.
```

Comment:
What does this second part mean, in practice? I'm having a hard time visualizing it.

Reply:
This is like when you apply a GraphQL selection set to a field that might return a list (in which case the selection set gets mapped over the list) but also might return a single object (in which case the selection set gets applied to just that object, no list involved).

So `actuallyNotAnArray->map(@.name)` still works, but you get the single string value `actuallyNotAnArray.name`, rather than array of strings.

Reply:
Makes sense, thanks!